### ISSUE TITLE
Limit nwm_client.gcp memory usage

### ISSUE THREAD

**Issue by jarq6c** [NOAA-OWP/hydrotools#127]
The default settings limit retrieval to channel reaches that coincide with USGS gage sites. However, things quickly spiral out of control when attempting to conduct large scale analyses that require all 2.7+ million NWM reaches. This submodule needs a way to limit the maximum number of values held in memory. This likely means a bit of a redesign and a change to the default cache.

**Comment by jarq6c** [comment]
I'll let this issue serve as a way to formalize requirements to efficiently retrieve and store model output for analysis.

Currently, we have a need to:
- [ ] Retrieve NetCDF output from Google Cloud Platform
- [ ] Retrieve NetCDF output from generic web servers (NOMADS, webserver directory listings, `http.server`)
- [ ] Generate `pandas.DataFrame`s containing time series of simulated or forecasted streamflow (in `hydrotools` canonical format)
- [ ] Generate `dask.dataframe`s containing time series of simulated or forecasted streamflow (in `hydrotools` canonical format)
- [ ] Work with a flexible generic `hydrotools.cache` interface
- [ ] Cache dataframes locally to HDF5 files in `table` format (via `hydrotools.cache`)
- [ ] Cache dataframes locally to parquet files (via `hydrotools.cache`)
- [ ] Take advantage of available compute resources to accelerate retrieval of NetCDF files and subsequent processing into dataframes
- [ ] The ability to crosswalk model output to arbitrary metadata based on `nwm_feature_id`

`pandas` and HDF5 covers light use-cases that are limited to USGS sites. `dask` and parquet cover heavier use-cases that may include all 2.7+ million NWM reaches.

The first step will be to refactor `caches` and `nwm_client` to use more generic interfaces based on abstract classes.

**Comment by jarq6c** [comment]
Tried various methods to go from remote non-DAP NetCDF files to a `pandas.DataFrame`. The best option for retrieving large amounts of data (on the order of 10<sup>4</sup> to 10<sup>6</sup> values) was to download the files locally and use `xarray.open_mfdataset` then `xarray.Dataset.to_dataframe`. This will likely circle us back to using `asyncio` to retrieve temporary files, merge data into a single dataframe, and then cache the resulting dataframe locally to HDF5, parquet, or zarr.

**Comment by aaraney** [comment]
I have a few comments that I will lump into one. Ill try to split it up so it is easily followable.

## Clarifying question

>  This submodule needs a way to limit the maximum number of values held in memory.

Should we limit the number of values held in memory or the memory amount? My vote would be amount of memory. I think it would be easier to enforce and implement. 

## Regarding the NetCDF in-memory problem

First off, this is interesting work that has use cases beyond the NWM. Just spitballing, it might be worth us looking around atmospheric data wrangling packages and see if/how they approach this problem. 

Regarding the NetCDF to `pandas.DataFrame` memory problem, my inclination is that HDF compression is actually being a hinderance in this case. Let me try to explain, to my knowledge, the NWM output uses HDF(4 or 5?) compression on top of the NetCDF format to further compress the data. In my limited knowledge of the HDF format, it uses a table and b-tree structures to store and account for data. Ergo, this often means a large number of small reads are done at the filesystem level to read data because the table describes exactly where in the tree to find the data. To my knowledge, the table and b-tree structures are not identical across NetCDF files even if two or more files contain the same coordinate and variable fields with the same number of records. That is, if the records are different (variables have different values between files). 

That being said, here is my naive and untested heuristic (hopefully there is already a software solution to this):
1. read HDF table from netcdf stream
2. determine portions of interest (im not sure the right HDF noun to use `group`, `superblock`, `dataset`?) 
3. read the stream of data only loading the portions desired in memory

Now, this sounds straightforward, but I doubt that will be the case.

**Comment by jarq6c** [comment]
Yes, the problem is amount of memory, not necessarily the number of values. The issue that arises has to do with how the current `gcp_client` retrieves files and constructs dataframes. The original assumption guiding `gcp_client` was targeted retrieval of streamflow simulations at limited locations (USGS sites). This meant we were generally retrieving between 1000 and 7000 features. The current `gcp_client`:

1. Retrieves NetCDF files in a single chunk per file and persists the file data in memory as a raw byte string
2. Reads the raw byte string into an `xarray.Dataset`
3. Filters the `Dataset` using a list of desired feature IDs
3. Converts the filtered `xarray.Dataset` to a `pandas.DataFrame`
4. Concatenates all the generated `pandas.DataFrames` into a single `DataFrame` and returns to the user

This process is memory intensive, but prioritizes speed by avoiding writing NetCDF files to disk. The most memory intensive portions are steps 2 and 4. These steps work fine on most compute resources, but things start to get out of hand as the list of feature IDs in step 3 gets larger than 100,000.

The solution I'm finding is to use the built-in methods that `xarray` and `dask` offer to read and process larger-than-memory datasets. An alternate solution (based on experimentation) might look like this:

1. Asynchronously retrieve many NetCDF files, streaming to disk in a temporary directory
2. Use `xarray` to lazily open the entire directory of files as a single `Dataset`
3. Convert the `Dataset` to a `dask.dataframe` or optimized `pandas.DataFrame` (user choice) and return to user

Memory limits can be enforced in step 1 and step 3. I've actually found this method to be pretty quick (at least on an SSD). Of note: the arrays inside an `xarray.Dataset` end up being `dask.array`, so the transformation to `dask.dataframe` is pretty seamless. This should reap benefits down the road when conducting large scale model evaluations due to the lazy loading, built-in parallelism, and `pandas` compatible interfaces offered by `dask`.
